apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs

```diff
@@ -60,21 +62,124 @@ private UrlSigner(IBlobSigner blobSigner, IClock clock)
+                await credential.GetDefaultServiceAccountEmailAsync(cancellationToken).ConfigureAwait(false),
+                CredentialsDefaultAlgorithm));
+
+        /// <summary>
+        /// Creates a new <see cref="UrlSigner"/> instance for a Google credential if
+        /// the wrapped credential type is supported for signing. In that case, this method
+        /// relies on the appropiate UrlSigner.FromCredential overload.
+        /// </summary>
+        /// <param name="credential">
+        /// The Google credential. Must not be null.
+        /// Must wrap a credential currently supported for signing.
+        /// </param>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <exception cref="InvalidOperationException">
+        /// The credential wrapped by the Google credential is not currently supported for signing.
+        /// </exception>
+        /// <remarks>
+        /// Please see the specific UrlSigner.FromCredential overloads for more information.
+        /// </remarks>
+        public static async Task<UrlSigner> FromCredentialAsync(GoogleCredential credential, CancellationToken cancellationToken = default) =>
+            GaxPreconditions.CheckNotNull(credential, nameof(credential)).UnderlyingCredential switch
+            {
+                ServiceAccountCredential sa => FromCredential(sa),
+                ImpersonatedCredential imp => FromCredential(imp),
+                ComputeCredential comp => await FromCredentialAsync(comp, cancellationToken).ConfigureAwait(false),
+                _ => throw new InvalidOperationException($"The credential type {credential.UnderlyingCredential.GetType()} is not supported for signing.")
+            };
+
+        /// <summary>
+        /// Creates a new <see cref="UrlSigner"/> instance from the JSON configuration file of a Google credential.
+        /// See <see cref="FromCredentialAsync(GoogleCredential, CancellationToken)"/> for more information about supported credential types.
+        /// </summary>
+        public static UrlSigner FromCredentialFilePath(string credentialFilePath) =>
+            // Compute credentials are not specified through configuration files, and that's the only
+            // case in which this call may be async. It's safe to just return result here.
+            FromCredentialAsync(GoogleCredential.FromFile(credentialFilePath)).ResultWithUnwrappedExceptions();
+
+        /// <summary>
+        /// Creates a new <see cref="UrlSigner"/> instance from the JSON configuration file of a Google credential,
+        /// which is read asyncrhonously.
+        /// See <see cref="FromCredentialAsync(GoogleCredential, CancellationToken)"/> for more information about supported credential types.
+        /// </summary>
+        public static async Task<UrlSigner> FromCredentialFilePathAsync(string credentialFilePath, CancellationToken cancellationToken = default) =>
```

Comment:
Possibly just FromCredentialFileAsync? That would be consistent with GoogleCredential.FromFileAsync.

I think it's fine for *this* method to be async, btw, given that the asynchronous part is loading the credential.